### DISCUSSION TITLE
EAP-MS-CHAPv2 generating NT-Response failed

### DISCUSSION THREAD

**Discussion by AlothNaveen** [strongswan/strongswan#2351] [Q&A]
[charon_024603.log.txt](https://github.com/user-attachments/files/16247805/charon_024603.log.txt)

**Comment by tobiasbrunner** [comment]
You might be missing a required algorithm (MD4, DES-ECB).

**Comment by AlothNaveen** [comment]
I am getting same error after addition of above plugins md4, des... I am unable to find DES-ECB plugin. Where to find it and how to add it.

strongswan_CHARON_PLUGINS := android-log socket-default openssl nonce oqs\
	pkcs1 pem x509 xcbc kdf revocation \
	eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls md4 des


~/strongswan-6.0.0beta6/src$ grep -r des-ecb
Binary file frontends/android/app/build/intermediates/stripped_native_libs/debug/out/lib/arm64-v8a/libcrypto.so matches
Binary file frontends/android/app/build/intermediates/stripped_native_libs/debug/out/lib/armeabi-v7a/libcrypto.so matches
Binary file frontends/android/app/build/intermediates/cxx/Debug/6nb676q2/obj/local/arm64-v8a/libcrypto.so matches
Binary file frontends/android/app/build/intermediates/cxx/Debug/6nb676q2/obj/local/armeabi-v7a/libcrypto.so matches
Binary file frontends/android/app/build/intermediates/merged_native_libs/debug/out/lib/arm64-v8a/libcrypto.so matches
Binary file frontends/android/app/build/intermediates/merged_native_libs/debug/out/lib/armeabi-v7a/libcrypto.so matches
frontends/android/app/src/main/jni/openssl/include/openssl/obj_mac.h:#define LN_des_ecb              "des-ecb"
Binary file frontends/android/app/src/main/jni/openssl/arm64-v8a/libcrypto.so matches
Binary file frontends/android/app/src/main/jni/openssl/armeabi-v7a/libcrypto.so matches
Binary file frontends/android/app/src/main/jni/openssl_111u_oqs/x86_64/libcrypto.a matches

**Reply by tobiasbrunner** [reply]
It could depend on the plugin order. If your version of OpenSSL provides one or more of the required algorithms, but then fails to actually calculate it successfully for some reason it won't make a difference if the other two plugins are loaded after the _openssl_ plugin. So either try to fix your OpenSSL version or move the two plugins to the front. (DES-ECB is provided by the _des_ plugin but as you can see probably also by OpenSSL.)

**Reply by AlothNaveen** [reply]
I changed open ssl shared library to static library. now i am getting received NO_PROPOSAL_CHOSEN notify error . The 1st error i got.
accept_private_algs = yes in strongswan.conf in rw strongsman vpn server side.
[charon_115731.log.txt](https://github.com/user-attachments/files/16278804/charon_115731.log.txt)
I followed #2344 link 
How to load oqs before charon ?

**Reply by tobiasbrunner** [reply]
> I changed open ssl shared library to static library.

Yes, that has to be statically linked, otherwise, the system library might get used. But you also load the other two plugins, which are technically not needed, before the _openssl_ plugin, so in regards to EAP-MSCHAPv2, the OpenSSL change probably won't make a difference.

> now i am getting received NO_PROPOSAL_CHOSEN notify error

Check the server log for details why that was returned.

> How to load oqs before charon ?

What do you mean? And why?

**Reply by AlothNaveen** [reply]
loaded plugins: androidbridge charon des md4 android-log socket-default openssl nonce oqs pkcs1 pem x509 xcbc kdf revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls  here can  load oqs before charon, may be charon is using  oqs

**Reply by AlothNaveen** [reply]
Still getting same error.
client log :
[charon_015615.log.txt](https://github.com/user-attachments/files/16280076/charon_015615.log.txt)
Server log: 
rsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) N(IKE_INT_SUP) ]

02[CFG] looking for an IKEv2 config for 192.168.1.100...192.168.1.101
02[CFG]   candidate: 192.168.1.100...%any, prio 1052
02[CFG] found matching ike config: 192.168.1.100...%any with prio 1052
02[IKE] local endpoint changed from 0.0.0.0[500] to 192.168.1.100[500]
02[IKE] remote endpoint changed from 0.0.0.0 to 192.168.1.101[39579]
02[IKE] 192.168.1.101 is initiating an IKE_SA
02[IKE] IKE_SA (unnamed)[3] state change: CREATED => CONNECTING
02[CFG] selecting proposal:
02[CFG]   no acceptable ENCRYPTION_ALGORITHM found
02[CFG] received proposals: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519/KE1_KYBER_L3
02[CFG] configured proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519/KE1_KYBER_L3
02[CFG] looking for IKEv2 configs for 192.168.1.100...192.168.1.101
02[CFG]   candidate: 192.168.1.100...%any, prio 1052
02[CFG] received supported signature hash algorithms: sha256 sha384 sha512 identity
02[IKE] remote host is behind NAT
02[IKE] sending strongSwan vendor ID
02[ENC] added payload of type VENDOR_ID to message
02[IKE] received proposals unacceptable
02[ENC] added payload of type NOTIFY to message
02[ENC] order payloads in message
02[ENC] added payload of type NOTIFY to message
02[ENC] generating IKE_SA_INIT response 0 [ N(NO_PROP) ]
02[ENC] not encrypting payloads
02[ENC] generating payload of type HEADER
02[ENC]   generating rule 0 IKE_SPI
02[ENC]   generating rule 1 IKE_SPI
02[ENC]   generating rule 2 U_INT_8
02[ENC]   generating rule 3 U_INT_4
02[ENC]   generating rule 4 U_INT_4
02[ENC]   generating rule 5 U_INT_8
02[ENC]   generating rule 6 RESERVED_BIT
02[ENC]   generating rule 7 RESERVED_BIT
02[ENC]   generating rule 8 FLAG
02[ENC]   generating rule 9 FLAG
02[ENC]   generating rule 10 FLAG
02[ENC]   generating rule 11 FLAG
02[ENC]   generating rule 12 FLAG
02[ENC]   generating rule 13 FLAG
02[ENC]   generating rule 14 U_INT_32
02[ENC]   generating rule 15 HEADER_LENGTH
02[ENC] generating HEADER payload finished
02[ENC] generating payload of type NOTIFY
02[ENC]   generating rule 0 U_INT_8
02[ENC]   generating rule 1 FLAG
02[ENC]   generating rule 2 RESERVED_BIT
02[ENC]   generating rule 3 RESERVED_BIT
02[ENC]   generating rule 4 RESERVED_BIT
02[ENC]   generating rule 5 RESERVED_BIT
02[ENC]   generating rule 6 RESERVED_BIT
02[ENC]   generating rule 7 RESERVED_BIT
02[ENC]   generating rule 8 RESERVED_BIT
02[ENC]   generating rule 9 PAYLOAD_LENGTH
02[ENC]   generating rule 10 U_INT_8
02[ENC]   generating rule 11 SPI_SIZE
02[ENC]   generating rule 12 U_INT_16
02[ENC]   generating rule 13 SPI
02[ENC]   generating rule 14 CHUNK_DATA
02[ENC] generating NOTIFY payload finished
02[NET] sending packet: from 192.168.1.100[500] to 192.168.1.101[39579] (36 bytes)
02[MGR] checkin and destroy IKE_SA (unnamed)[3]
02[IKE] IKE_SA (unnamed)[3] state change: CONNECTING => DESTROYING
04[NET] sending packet: from 192.168.1.100[500] to 192.168.1.101[39579]
02[MGR] checkin and destroy of IKE_SA successful

**Reply by tobiasbrunner** [reply]
> loaded plugins: androidbridge charon des md4 android-log socket-default openssl nonce oqs pkcs1 pem x509 xcbc kdf revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls here can load oqs before charon, may be charon is using oqs

That's not how this works. It's perfectly fine if _oqs_ is at the end, it only provides PQC algorithms that no other plugin provides anyway.



> Still getting same error.

The log tells you exactly what the problem is: 

    02[CFG] no acceptable ENCRYPTION_ALGORITHM found
